[2.9 updates] New custom XCom backend content (#3566)
astronomer/docs: learn/airflow-custom-xcom-backends.md

---
title: 'Options for custom XCom backends in Airflow'
sidebar_label: 'Custom XCom backends'
id: airflow-custom-xcom-backends
description: 'Use this guide to learn about different ways to set up custom XCom backends.'
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> Airflow [XComs](airflow-passing-data-between-tasks.md) allow you to pass data between tasks. By default, Airflow uses the [metadata database](airflow-database.md) to store XComs, which works well for local development but has limited performance. With **Airflow custom XCom backends** you can configure where and how Airflow stores XComs, and customize serialization and deserialization methods.

Reviewer: ```suggestion
Airflow [XComs](airflow-passing-data-between-tasks.md) allow you to pass data between tasks. By default, Airflow uses the [metadata database](airflow-database.md) to store XComs, which works well for local development but has limited performance. If you configure a custom XCom backend, you can define where and how Airflow stores XComs, as well as customize serialization and deserialization methods.
```

In this guide you'll learn:

> - Why you might want to use a custom XCom backend.

Reviewer: ```suggestion
- When you should use a custom XCom backend.
```

- How to set up a custom XCom backend using the Object Storage custom XCom backend.
- How to use a custom XCom backend class with custom serialization and deserialization methods.
- How to override the `.clear()` method to delete XComs from a custom XCom backend when clearing tasks.

:::warning

While a custom XCom backend allows you to store virtually unlimited amounts of data as XComs, you will also need to scale other Airflow components to pass large amounts of data between tasks. For help running Airflow at scale, [reach out to Astronomer](https://www.astronomer.io/try-astro/?referral=docs-content-link&utm_medium=docs&utm_content=learn-xcom-backend-tutorial&utm_source=body).

:::

## Assumed Knowledge

To get the most benefits from this guide, you need an understanding of:

- XCom basics. See [Pass data between tasks](airflow-passing-data-between-tasks.md).
- Basic knowledge of a cloud-based object storage service like [AWS S3](https://aws.amazon.com/s3/), [GCP Cloud Storage](https://cloud.google.com/storage) or [Azure Blob Storage](https://azure.microsoft.com/en-us/products/storage/blobs/).

## Why use a custom XCom backend?

> Common reasons to use a custom XCom backend include:

Reviewer: Do we recommend a custom backend for all prod workflows? It might be nice to have a single "hero" reason before this list even starts.

> - Needing more storage space for XComs than the Airflow metadata database can offer.
> - Running a production environment where you require custom retention, deletion, and backup policies for XComs.
> - Accessing XComs without accessing the metadata database.
> - Restricting types of allowed XCom values.
> - Saving XComs in multiple locations simultaneously.

Reviewer: ```suggestion
- You need more storage space for XComs than the Airflow metadata database can offer.
- You're running a production environment where you require custom retention, deletion, and backup policies for XComs.
- You want to access XComs without accessing the metadata database.
- You want to restrict types of allowed XCom values.
- You want to save XComs in multiple locations simultaneously.
```

> It is also possible to use custom XCom backends to define custom serialization and deserialization methods for XComs if adding a serialization method to a class or registering a custom serializer is not feasible. See [Custom serialization and deserialization](#custom-serialization-and-deserialization) for more information.

Reviewer: ```suggestion
You can also use custom XCom backends to define custom serialization and deserialization methods for XComs if you need to add a serialization method to a class, or if registering a custom serializer is not feasible. See [Custom serialization and deserialization](#custom-serialization-and-deserialization) for more information.
```

## How to set up a custom XCom backend

There are two main ways to set up a custom XCom backend:

> - **Object Storage custom XCom backend**: Use this method to create a custom XCom backend when you want to store XComs in a cloud-based object storage service like AWS S3, GCP Cloud Storage, or Azure Blob Storage.

Reviewer: Should we say "this option is recommended if you have no specific requirements beyond needing more XCom storage" or something similar?

> - **Custom XCom backend class**: Use this method when you want to further customize how XComs are stored, for example simultaneously storing XComs in two different locations.

Reviewer: ```suggestion
- **Custom XCom backend class**: Use this method when you want to further customize how XComs are stored, for example to simultaneously store XComs in two different locations.
```

> Additionally, some provider packages offer custom XCom backends that you can use out of the box, for example the [Snowpark provider](airflow-snowpark.md) which contains a custom XCom backend for Snowflake.

Reviewer: ```suggestion
Additionally, some provider packages offer custom XCom backends that you can use out of the box. For example, the [Snowpark provider](airflow-snowpark.md) contains a custom XCom backend for Snowflake.
```

> ### Use the Object Storage XCom backend

Reviewer: ```suggestion
### Use the Object Storage XCom Backend
```
Is this the official name of the feature? Maybe there's another way to describe it, but that's the impression I'm getting

Author: Yes, https://airflow.apache.org/docs/apache-airflow-providers-common-io/stable/xcom_backend.html :)

> Airflow 2.9+ added the possibility to create a custom XCom backend using the Object Storage feature. Object Storage custom XCom backends are part of the [Common IO](https://registry.astronomer.io/providers/apache-airflow-providers-common-io/versions/latest) provider and can be defined only using environment variables. The following environment variables are available:

Reviewer: ```suggestion
Airflow 2.9+ added the possibility to create a custom XCom backend using object storage. The Object Storage XCom Backend is part of the [Common IO](https://registry.astronomer.io/providers/apache-airflow-providers-common-io/versions/latest) provider and can be defined using the following environment variables:
```

        query = query.filter_by(map_index=map_index)

    # iterate through xcoms of this task and delete them from S3
    xcom_entries = query.all()
    for xcom_entry in xcom_entries:
        reference_string = xcom_entry.value
        if reference_string:
            print(reference_string)
            key = reference_string.replace(MyCustomXComBackend.PREFIX, "")
            hook = S3Hook(aws_conn_id="my_aws_conn")
            hook.delete_objects(
                bucket=MyCustomXComBackend.S3_BUCKET_NAME, keys=[key]
            )

    # delete the XCom containing the reference string from metadata database
    query.delete()
```

</div>
</details>

## Custom serialization and deserialization

By default, Airflow includes [serialization](https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/serializers.html) methods for common object types like [JSON](https://www.json.org/json-en.html), [pandas DataFrames](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) and [NumPy](https://numpy.org/). You can see all supported types in the [Airflow source code](https://github.com/apache/airflow/tree/main/airflow/serialization/serializers).

> In cases where you need to pass data objects through XCom that are not supported you have several options:

Reviewer: ```suggestion
If you need to pass data objects through XCom that are not supported, you have several options:
```
